File: core/src/de/sharedinbox/core/repository/MailboxRepository.kt
package de.sharedinbox.core.repository

import de.sharedinbox.core.jmap.mail.Mailbox
import kotlinx.coroutines.flow.Flow

interface MailboxRepository {
    /** Emits the mailbox list for [accountId] from the local DB, re-emits on sync. */
    fun observeMailboxes(accountId: String): Flow<List<Mailbox>>

    /** Runs Mailbox/get + Mailbox/changes against the server and writes results to DB. */
    suspend fun syncMailboxes(accountId: String): Result<Unit>

    /**
     * Creates a new mailbox on the server and stores it locally.
     *
     * Returns [Result.failure] with a descriptive message when the server rejects the
     * creation (e.g. due to insufficient permissions). The failure is also recorded in
     * the sync log so the user can review it.
     */
    suspend fun createMailbox(accountId: String, name: String, parentId: String? = null): Result<Mailbox>

    /**
     * Renames [mailboxId] to [newName] on the server and updates the local DB.
     *
     * Conflicts (server rejection) are logged and surfaced as [Result.failure].
     */
    suspend fun renameMailbox(accountId: String, mailboxId: String, newName: String): Result<Unit>

    /**
     * Deletes [mailboxId] on the server and removes it from the local DB.
     *
     * Conflicts (server rejection, e.g. non-empty mailbox) are logged and surfaced
     * as [Result.failure].
     */
    suspend fun deleteMailbox(accountId: String, mailboxId: String): Result<Unit>
}

File: lib/core/repositories/mailbox_repository.dart
import 'package:sharedinbox/core/models/mailbox.dart';

abstract class MailboxRepository {
  Stream<List<Mailbox>> observeMailboxes(String? accountId);

  /// Returns the number of mailboxes synced.
  Future<int> syncMailboxes(String accountId);

  /// Returns the first mailbox with the given [role] for [accountId], or null.
  Future<Mailbox?> findMailboxByRole(String accountId, String role);

  /// Deletes all locally-cached mailbox rows for [accountId].
  Future<void> clearForResync(String accountId);

  /// Creates a new mailbox named [name] for [accountId] and tags it with
  /// [role] in the local database. For JMAP accounts the role is also sent
  /// to the server. Returns the newly created [Mailbox].
  ///
  /// When [parentDisplayPath] is non-null, the new mailbox is created as a
  /// child of the mailbox with that [Mailbox.displayPath] in the local cache.
  Future<Mailbox> createMailboxWithRole(
    String accountId,
    String name,
    String role, {
    String? parentDisplayPath,
  });

  /// Creates a new mailbox named [name] for [accountId] without a special role.
  /// Returns the newly created [Mailbox].
  ///
  /// When [parentDisplayPath] is non-null, the new mailbox is created as a
  /// child of the mailbox with that [Mailbox.displayPath] in the local cache.
  /// The mailbox is created on the account's IMAP/JMAP server, so this call
  /// requires network connectivity.
  Future<Mailbox> createMailbox(
    String accountId,
    String name, {
    String? parentDisplayPath,
  });

  /// Renames the mailbox with [mailboxPath] under [accountId] so that its
  /// leaf name becomes [newName]. Parent stays the same. Returns the updated
  /// [Mailbox]. Talks to the account's IMAP/JMAP server.
  Future<Mailbox> renameMailbox(
    String accountId,
    String mailboxPath,
    String newName,
  );

  /// Deletes the mailbox with [mailboxPath] under [accountId] on the server
  /// and removes its local cache (emails, threads, sync state).
  Future<void> deleteMailbox(String accountId, String mailboxPath);

  /// Moves the mailbox with [mailboxPath] under [accountId] to live under
  /// [newParentDisplayPath] (null for top level). The leaf name is preserved.
  /// Returns the updated [Mailbox].
  Future<Mailbox> moveMailbox(
    String accountId,
    String mailboxPath, {
    required String? newParentDisplayPath,
  });
}
